soc: arm: rpi_pico: Add support for binary info feature

soburi opened the pull request:
soc: arm: rpi_pico: Add support RaspberryPi Pico binary info feature

Binary Info embeds program meta information in flash,
which can be viewed with RaspberryPi Pico's `picotool`.

Metadata is automatically collected from pinctrl.

It can be override by the Kconfig configurations, such as

```
CONFIG_RPI_PICO_BINARY_INFO_OVERRIDE_PROGRAM_NAME="my program name"
```

When this feature is enabled, pinctrl groups are restricted to
consisting of pins with a single rpi_pico function.
In other words, SPI's MISO and MOSI can be in the same group,
but I2C's SDA cannot be included in this group.
This is rarely an issue in normal use,
and can be resolved by dividing them into separate groups.

soburi committed:
soc: raspberrypi: rpi_pico: Add support RpiPico binary info feature

Binary Info embeds program meta information in flash,
which can be viewed with RaspberryPi Pico's `picotool`.

Metadata is automatically collected from pinctrl.

It can be override by the Kconfig configurations, such as

```
CONFIG_RPI_PICO_BINARY_INFO_OVERRIDE_PROGRAM_NAME="my program name"
```

When this feature is enabled, pinctrl groups are restricted to
consisting of pins with a single rpi_pico function.
In other words, SPI's MISO and MOSI can be in the same group,
but I2C's SDA cannot be included in this group.
This is rarely an issue in normal use,
and can be resolved by dividing them into separate groups.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

mbolivar-nordic reviewed (commented):
Thank you for your work!

I am not sure about the CMake variable changes here. I'd like @tejlmand to confirm it is OK.

yonsch commented:
Before I comment on the changes themselves - I think this feature would be beneficial for all Zephyr users, not only for the pico. Being able to embed strings in a binary where they're accessable outside of the image is very useful. The main use case I can think of is a bootloader and an app being able to read each other's versions. Since this would require some more work I'm not sure it should block this PR for the meantime, but that is something to keep in mind.
I should also note that pico's binary_info can be used to store pin configurations, but I can't think of a use case for that.

soburi commented on soc/arm/rpi_pico/rp2/binary_info.c:
```diff
@@ -0,0 +1,119 @@
+bi_decl(bi_program_name((uint32_t)PICO_PROGRAM_NAME));
+#endif
+
+#ifdef PICO_BOARD
```
There are no strong reason to configure the settings by CMake variables. Removed it.
(It is port from Pico-SDK. It may sometimes convenient, but no strong reason to keep it.)

mbolivar-nordic replied:
Thanks, resolving

soburi commented:
@yonsch 
> Before I comment on the changes themselves - I think this feature would be beneficial for all Zephyr users, not only for the pico. Being able to embed strings in a binary where they're accessable outside of the image is very useful. The main use case I can think of is a bootloader and an app being able to read each other's versions. Since this would require some more work I'm not sure it should block this PR for the meantime, but that is something to keep in mind. I should also note that pico's binary_info can be used to store pin configurations, but I can't think of a use case for that.

The 'self-reference'-ish feature is interresting suggestion.
But I also think there are some difficult point. 
In the Pico's scope, the information mainly read it by BOOTSEL mode.
This is probably the same for other MCUs, and I think this function is strongly related to the "recovery mode" specification.
If there are other MCUs with similar functions, it may be possible to define an abstract interface.
If I can draw a big picture, I think it would wonderful if it could define as uf2 meta information. (Create a .uf2meta section with a linker and have the same information there. It can handle with common specification by recovery program and firmware itself!)

yonsch commented:
@soburi I've created a quick demonstration of what I was thinking of: #54464

MaureenHelm commented:
dev-review: depends on #54464

soburi commented:
@cfriedt 

> I assume that once #54464 is merged, work on this will continue

That's right. Perhaps I will have to rethink the design accordingly.

yonsch commented:
@soburi Why does this depend on #54464?
Bindesc was supposed to be a standard Zephyr replacement for binary info.
Binary info would still be useful for someone who needs to specifically use the pico format, e.g. if a bootloader expects binary info descriptors.
I think we can treat them as if they are mutually exclusive, as they serve the same purpose, and proceed with this PR regardless of bindesc.
Unless, you somehow see them coexisting somehow, which would require more planning.

MaureenHelm commented:
> @cfriedt
> 
> > I assume that once #54464 is merged, work on this will continue
> 
> That's right. Perhaps I will have to rethink the design accordingly.

Has this been superseded by #54464? Can we close?

soburi committed:
modules: hal_rpi_pico: add includes that referenced by binary info

The binary info feature references `boot_stage2` and `pico_binary_info`.
Add these to include dirs.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
modules: hal_rpi_pico: Make PICO_NO_BINARY_INFO configurable

Use the definition of PICO_NO_BINARY_INFO if it already exists,
set it as 1 if not so.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

dsseng commented on soc/raspberrypi/rpi_pico/common/binary_info_header.S:
Could this potentially be handled by a linker script like rom_start?

soburi committed:
soc: raspberrypi: rpi_pico: Fix indentation

Fix indentation to pass CI check.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi commented on soc/raspberrypi/rpi_pico/common/binary_info_pins.c:
```diff
@@ -0,0 +1,1109 @@
+BINARY_INFO_FROM_PINCFG_GROUP(DT_NODELABEL(pinctrl), 31, 29);
+BINARY_INFO_FROM_PINCFG_GROUP(DT_NODELABEL(pinctrl), 31, 30);
+BINARY_INFO_FROM_PINCFG_GROUP(DT_NODELABEL(pinctrl), 31, 31);
+#endif
```
We can't use anything other than LINE, and I don't have the habit of writing code snippets.
There's no other way.

dsseng reviewed (approved):
It's alright, not that busy :). Looks great from my perspective!
